[BE] Job comments

nspc97 opened the pull request:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

Users should have the possibility to leave comments for jobs. Replies should be possible.

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

Users should have the possibility to leave comments for jobs. Replies should be possible.

SpartakusMd commented on src/Entity/Comment.php:
```diff
@@ -0,0 +1,169 @@
+     * @Groups({"Comment"})
+     * @SWG\Property()
+     */
+    private $reply;
```
Is this the id of the comment to which you reply or the id of the reply one?

Anyway, this should be a relation OneToMany self referenced https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/association-mapping.html#one-to-many-self-referencing

SpartakusMd replied:
You didn't answer the question. As the name creates uncertainty, please rename it to `$parent`.

SpartakusMd commented on config/packages/nelmio_api_doc.yaml:
```diff
@@ -12,11 +12,14 @@ nelmio_api_doc:
       - { alias: UserList, type: App\Entity\User, groups: [UserList] }
       - { alias: UserForActivityDetails, type: App\Entity\User, groups: [ActivityDetails] }
       - { alias: UserForActivityCreate, type: App\Entity\User, groups: [ActivityCreate] }
+      - { alias: UserForComment, type: App\Entity\User, groups: [Comment] }
```
Keep the same format as always. Should be `CommentList`.

SpartakusMd commented on src/Repository/CommentRepository.php:
```diff
@@ -0,0 +1,50 @@
+        $em->flush();
+    }
+
+    public function getCommentsFotActivity(Activity $activity): QueryBuilder
```
You have a typo `getCommentsForActivity`.

SpartakusMd commented on src/Entity/Comment.php:
```diff
@@ -0,0 +1,169 @@
+        return $this->comment;
+    }
+
+    public function setComment(string $comment): self
```
Why some setters `return $this;` but some don't? Please be consistent.

SpartakusMd commented on src/Transformer/ActivityTransformer.php:
```diff
@@ -141,4 +143,31 @@ public function editTransform(
+        $entity->setActivity($activity);
+        $entity->setUser($user);
+        $entity->setComment($commentDTO->comment);
+        $entity->setReply($commentDTO->reply);
```
Check the existence of `$commentDTO->reply`.

MihaiDruta commented on src/Controller/ActivityController.php:
```diff
@@ -1446,4 +1449,293 @@ public function declineInvitation(
+     *     @SWG\Property(property="message", type="string", example="Your comment has successfully added!"),
+     *     )
+     * )
+     * @SWG\Response(
```
Response code 403 is never shown in a response when adding a comment. Maybe a validation is needed (you shouldn't be able to comment on a unavailable/private activity)

SpartakusMd replied:
Nice catch 👍

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

Migration

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

Added MaxDepth constraint to SerializationContext

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

SpartakusMd commented on src/Entity/Comment.php:
```diff
@@ -0,0 +1,178 @@
+{
+    /**
+     * @ORM\Id()
+     * @OneToMany(targetEntity="Comment", mappedBy="reply")
```
Why make the relation bedirectional?

SpartakusMd commented on src/Entity/Comment.php:
```diff
@@ -0,0 +1,178 @@
+    /**
+     * User id.
+     * @ORM\ManyToOne(targetEntity="User")
+     * @ORM\JoinColumn(onDelete="CASCADE")
```
This will create problems. Test case to try:
1. Create a comment with user1
2. Reply to that comment with user2
3. Remove user1

SpartakusMd commented on src/Controller/ActivityController.php:
```diff
@@ -1446,4 +1449,303 @@ public function declineInvitation(
+     * @throws ORMException
+     * @throws OptimisticLockException
+     */
+    public function addComment(
```
Move the comment endpoints to a separate controller `CommentController`.

SpartakusMd commented on src/Transformer/ActivityTransformer.php:
```diff
@@ -141,4 +150,34 @@ public function editTransform(
+     * @param User $user
+     * @return Comment
+     */
+    public function addComment(CommentDTO $commentDTO, Activity $activity, User $user): Comment
```
Move addComment & editComment to CommentTransformer.

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

- Moved addComment & editComment to CommentTransformer.
- Moved the comment endpoints to a separate controller CommentController.
- When deleting an user his comments remain actual with user_id equal null
- Removed the bidirectional relation between id - parent

SpartakusMd commented on src/Controller/ActivityController.php:
```diff
@@ -3,8 +3,10 @@
 namespace App\Controller;
 
 use App\DTO\ActivityDTO;
+use App\DTO\CommentDTO;
```
Not used.

SpartakusMd commented on src/Controller/ActivityController.php:
```diff
@@ -3,8 +3,10 @@
+use App\DTO\CommentDTO;
 use App\Entity\Activity;
 use App\Entity\ActivityUser;
+use App\Entity\Comment;
```
Not used.

SpartakusMd commented on src/Controller/ActivityController.php:
```diff
@@ -15,6 +17,7 @@
 use App\Handlers\ActivityHandler;
 use App\Handlers\ActivityUserHandler;
 use App\Repository\ActivityUserRepository;
+use App\Repository\CommentRepository;
```
Not used.

SpartakusMd commented on src/Controller/CommentController.php:
```diff
@@ -0,0 +1,358 @@
+
+    /**
+     * Edit comment.
+     * @Rest\Post("/activities/{activityId}/edit_comment/{commentId}", requirements={"id"="\d+"})
```
In requirements, you have `id` which is not present in query. In query you have `activityId` and `commentId` which are not present in requirements.

SpartakusMd commented on src/Transformer/CommentTransformer.php:
```diff
@@ -0,0 +1,60 @@
+        return $comment;
+    }
+
+    public function clearUserComments(User $user): void
```
This should not be part of Transformer. You are not transforming anything here, right? It should be in the CommentRepository.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -317,6 +324,7 @@ public function deleteUser(User $user, UserRepository $userRepository): JsonResp
                 'message' => 'Access denied!'
             ], Response::HTTP_FORBIDDEN);
         }
+        $this->commentTransformer->clearUserComments($authenticatedUser);
```
This should be called by `$userRepository->delete()`. Like that you won't forget about clearing user comments in other places.

SpartakusMd commented on src/Entity/Comment.php:
```diff
@@ -0,0 +1,175 @@
+
+    /**
+     * Comment id linked to.
+     * @ORM\ManyToOne(targetEntity="Comment", inversedBy="id")
```
`inversedBy` is used only for bidirectional relations.

SpartakusMd commented on src/Transformer/CommentTransformer.php:
```diff
@@ -0,0 +1,60 @@
+        $entity->setUser($user);
+        $entity->setComment($commentDTO->comment);
+        if ($commentDTO->parent) {
+            $parentComment = $this->commentRepository->find($commentDTO->parent);
```
What will happen if you don't find parent id? Shouldn't you throw an error?

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

SpartakusMd commented on src/Controller/CommentController.php:
```diff
@@ -166,14 +167,27 @@ public function addComment(
+        } catch (EntityNotFound $exception) {
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_NOT_FOUND,
```
Use bad request or precondition failed.

SpartakusMd commented on src/Repository/CommentRepository.php:
```diff
@@ -47,4 +48,12 @@ public function getCommentsForActivity(Activity $activity): QueryBuilder
+        $userComments = $this->findBy(array('user' => $user));
+        foreach ($userComments as $userComment) {
+            $userComment->setUser(null);
+        }
```
You forgot to flush the changes so they arrinve in DB.

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments

SpartakusMd commented on src/Repository/CommentRepository.php:
```diff
@@ -49,11 +49,17 @@ public function getCommentsForActivity(Activity $activity): QueryBuilder
         $userComments = $this->findBy(array('user' => $user));
         foreach ($userComments as $userComment) {
             $userComment->setUser(null);
+            $this->save($userComment);
```
Do an `$em->flush();` outside of foreach (instead of this save) so that all the updates are done via one query to the database.

Commit:
[BE] Job comments
https://trello.com/c/vRfRn8EZ/85-be-job-comments
